sdk/eventhub/event-hubs/src/eventHubProducerClient.ts

```diff
@@ -224,32 +253,65 @@ export class EventHubProducerClient {
           `The partitionId (${unexpectedOptions.partitionId}) set on sendBatch does not match the partitionId (${partitionId}) set when creating the batch.`
         );
       }
+
+      spanContextsToLink = batch._messageSpanContexts;
     } else {
+      if (!Array.isArray(batch)) {
+        batch = [batch];
+      }
+
       // For arrays of events, partitionId and partitionKey would be set in the options.
       const expectedOptions = options as SendBatchOptions;
       partitionId = expectedOptions.partitionId;
       partitionKey = expectedOptions.partitionKey;
+
+      for (let i = 0; i < batch.length; i++) {
+        const event = batch[i];
+        if (!event.properties || !event.properties[TRACEPARENT_PROPERTY]) {
+          const messageSpan = createMessageSpan(getParentSpan(options.tracingOptions));
+          // since these message spans are created from same context as the send span,
+          // these message spans don't need to be linked.
+          // replace the original event with the instrumented one
+          batch[i] = instrumentEventData(batch[i], messageSpan);
+          messageSpan.end();
+        }
+      }
     }
     if (partitionId && partitionKey) {
       throw new Error(
         `The partitionId (${partitionId}) and partitionKey (${partitionKey}) cannot both be specified.`
       );
     }
 
-    if (!partitionId) {
-      // The producer map requires that partitionId be a string.
-      partitionId = "";
+    let sender = this._sendersMap.get(partitionId || "");
```

Comment:
I just realized...should we cast the user-provided `partitionId` to a string when we assign to the local partitionId? I think for JS users we've allowed them to specify partitionId as a number when receiving events, and in this case if they specified partition '0' as a number, it'd be treated as if it didn't exist in a few of our checks.

Reply:
Yes, we should cast it to string. This also opened quite a few places where we are not doing this when sending. I'll send a fix later in the day